olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) 2019 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#pragma once
+
+#include <olp/core/client/ApiResponse.h>
+#include <olp/core/client/CancellationContext.h>
+
+namespace olp {
+namespace thread {
+
+/**
+ * @brief The TaskScheduler class is an abstract interface to be used as base
+ * for a custom thread scheduling strategy.
+ *
+ * Subclasses should inherit from this class and implement virtual EnqueueTask,
+ * that takes any callable target (lambda expression, bind expression or any
+ * other function object) as input and adds it to the execution pipeline.
+ *
+ */
+class CORE_API TaskScheduler {
+ public:
+  /// Alias for abstract interface input
+  using CallFuncType = std::function<void()>;
+
+  virtual ~TaskScheduler() = default;
```

Comment:
Put destructor definition into separate .cpp . This way you can also mention it in CMakeLists and make sure it compiles

Reply:
As this will be used pretty soon by the actual impl I assume this check will be soon performed. No need to place extra cpp just for a default destructor.

Reply:
```
[  4%] Building CXX object olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/src/thread/TaskScheduler.cpp.o
In file included from /mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/src/thread/TaskScheduler.cpp:1:
/mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h:62:12: error: 'this' argument to member function 'IsCancelled' has type 'const client::CancellationContext', but function is not marked const
      if (!context.IsCancelled()) {
           ^~~~~~~
/mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/include/olp/core/client/CancellationContext.h:74:15: note: 'IsCancelled' declared here
  inline bool IsCancelled() {
              ^
In file included from /mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/src/thread/TaskScheduler.cpp:1:
/mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h:61:24: error: call to implicitly-deleted copy constructor of 'client::CancellationContext'
    auto task = [func, context]() {
                       ^~~~~~~
/mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/include/olp/core/client/CancellationContext.h:80:24: note: copy constructor of 'CancellationContext' is implicitly deleted because field 'mutex_' has a deleted copy constructor
  std::recursive_mutex mutex_;
                       ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/mutex:101:5: note: 'recursive_mutex' has been explicitly marked deleted here
    recursive_mutex(const recursive_mutex&) = delete;
    ^
In file included from /mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/src/thread/TaskScheduler.cpp:1:
/mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h:67:12: error: call to implicitly-deleted copy constructor of 'client::CancellationContext'
    return context;
           ^~~~~~~
/mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/include/olp/core/client/CancellationContext.h:80:24: note: copy constructor of 'CancellationContext' is implicitly deleted because field 'mutex_' has a deleted copy constructor
  std::recursive_mutex mutex_;
                       ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/mutex:101:5: note: 'recursive_mutex' has been explicitly marked deleted here
    recursive_mutex(const recursive_mutex&) = delete;
    ^
3 errors generated.
olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/build.make:974: recipe for target 'olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/src/thread/TaskScheduler.cpp.o' failed
make[3]: *** [olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/src/thread/TaskScheduler.cpp.o] Error 1
CMakeFiles/Makefile2:503: recipe for target 'olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/all' failed
make[2]: *** [olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/all] Error 2
CMakeFiles/Makefile2:515: recipe for target 'olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/rule' failed
make[1]: *** [olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/rule] Error 2
Makefile:255: recipe for target 'olp-cpp-sdk-core' failed
make: *** [olp-cpp-sdk-core] Error 2
```

Reply:
Yes, as this https://github.com/heremaps/here-olp-edge-sdk-cpp/pull/37 is missing.